### ISSUE TITLE
Rule proposal: no-labeled-statement

### ISSUE THREAD

**Issue by jfmengels** [sindresorhus/eslint-plugin-unicorn#69]
In JavaScript, there is a little known feature called [labels](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Statements/label), which allows to do the equivalent of gotos.

The reason I want to forbid this is two-fold:
1. gotos are bad, and they should not be used
2. But mostly, it has become really easy to write these labeled statements instead of writing objects, with the introduction of arrow functions.

``` js
const foo = value => { foo: value };
```

The previous excerpt does not return an object with a key `foo`, but returns a [code block that will be executed and not do much](https://astexplorer.net/#/gFvmvVsIKD).

The following should have been used instead, where we explicitly return an expression (due to the parentheses) that contains an object expression:

``` js
const foo = value => ({ foo: value });
```

I propose to write a rule, named samething like either `no-labeled-statement` or `no-confusing-arrow-label` (hint to the existing [no-confusing-arrow](http://eslint.org/docs/rules/no-confusing-arrow)) where the error message would warn against using labels in arrow functions, and maybe advising to wrap the returned block. Depending on the name we choose, this will forbid labeled statements altogether (they are bad anyway) or just those underneath an arrow function.

Unfortunately, this rule would still not be able to help with the following...

``` js
const foo = value => { value };
```

PS: I think we should have a GitHub label for rule proposals yet to be accepted. Sounds good?

**Comment by sindresorhus** [comment]
ESLint already has a built-in rule to prevent labels in general: [`no-labels`](http://eslint.org/docs/rules/no-labels). Would be natural to add an option there to also include labels in arrow functions. What do you think?

-

> PS: I think we should have a GitHub label for rule proposals yet to be accepted. Sounds good?

👍 Label name suggestion?

**Comment by jfmengels** [comment]
> no-labels. Would be natural to add an option there to also include labels in arrow functions. What do you think?

Sounds good. I'm pretty amazed that it doesn't forbid it altogether with such a name. Putting it in my todo list.

> Label name suggestion?

I like `evaluating`, which ESLint uses.

**Comment by futpib** [comment]
But `no-labels` does ban labels like that, and xo does report `no-labels` violations as errors with default setup.

I just checked it like this:
```bash
echo 'const foo = value => { foo: value };' | yarn xo --stdin
```
